isisd/isis_tlvs.c

```diff
@@ -3467,10 +4387,22 @@ static void tlvs_protocols_supported_to_adj(struct isis_tlvs *tlvs,
 	memcpy(adj->nlpids.nlpids, reduced.nlpids, reduced.count);
 }
 
+DEFINE_HOOK(isis_adj_ipv4_enabled_hook, (struct isis_adjacency *adj), (adj))
+DEFINE_HOOK(isis_adj_ipv4_disabled_hook, (struct isis_adjacency *adj), (adj))
+DEFINE_HOOK(isis_adj_ipv6_enabled_hook, (struct isis_adjacency *adj), (adj))
+DEFINE_HOOK(isis_adj_ipv6_disabled_hook, (struct isis_adjacency *adj), (adj))
+
 static void tlvs_ipv4_addresses_to_adj(struct isis_tlvs *tlvs,
 				       struct isis_adjacency *adj,
 				       bool *changed)
 {
+	bool ipv4_enabled = false;
+
+	if (adj->ipv4_address_count == 0 && tlvs->ipv4_address.count > 0)
+		ipv4_enabled = true;
+	else if (adj->ipv4_address_count > 0 && tlvs->ipv4_address.count == 0)
+		hook_call(isis_adj_ipv4_disabled_hook, adj);
+
 	if (adj->ipv4_address_count != tlvs->ipv4_address.count) {
```

Comment:
The `else if` is never executed when adjacency is removed due to a bug in isisd code. 

If `isis_adj_ipv4_XXX_hook()` is merged with IPv6 one, it is easier to add AF_INET parameter here.

Why not using a direct call to `isis_sr_adj_ipv4_enabled()` function instead of using hook_call() ?

Reply:
This hook will be called only when the adjacency runs out of IPv4 addresses. When the adjacency goes down altogether, the `isis_adj_state_change_hook` hook is called instead. The SR code registers a callback to both hooks to handle these different events.

> Why not using a direct call to `isis_sr_adj_ipv4_enabled()` function instead of using hook_call() ?

Mostly to make the code more modular. In the future we might want to register other callbacks to this hook.